sklearn/utils/openmp_helpers.pyx

```diff
@@ -0,0 +1,41 @@
+IF SKLEARN_OPENMP_SUPPORTED:
+    cimport openmp
+    from joblib import cpu_count
+
+
+cpdef _openmp_effective_n_threads(n_threads=None):
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
```

Comment:
We should always pass `n_threads=None` in the code right? The rest should only be for debugging?

Reply:
Most of the time yes.

Few exceptions : if we keep n_jobs in kmeans we would pass it to this function. If we decide to add a parameter to histgradientbosstingtrees to control the number of threads we would pass it to.

Reply:
wait, we agreed in https://github.com/scikit-learn/scikit-learn/issues/14265 not to expose a parameter for openmp threads. We should deprecate `n_jobs` for Kmeans and link to the doc from #15116 now right?

Reply:
The openmp implement of kmeans is not merged yet so it's not yet needed.

We agreed to do that case by case. For most situations we agreed to expose nothing but we also agreed that we can discuss it for estimators where the main loop is parallelized with openmp.

Reply:
What do you think of `sklearn.set_config(n_omp_threads=...)`, which can be used directly in `prange(..., num_threads)`? The highest python level caller could get the value and pass it down into the cython function.

Reply:
> We agreed to do that case by case

Really? Where? My understanding is that we went for option 5.

Going case by case (i.e. accepting n_jobs for controlling openMP in some cases) is going to be extremely painful to explain to users.  #15116  is already a lot to digest. We need to keep things simple

Reply:
> What do you think of sklearn.set_config(n_omp_threads=...),

I think that we agreed on 5 and we should stick with 5. This is not the time for new ideas... That time was for #15116 

(but to address your solution, I don't see how this is different from what setting OMP_NUM_THREADS)

Reply:
> Really? Where? My understanding is that we went for option 5.

during the last meeting. We went for option 5. But we allowed to keep the discussion opened on a case by case. It does not mean we won't use 5 everywhere :)